### DISCUSSION TITLE
How to use keycloak-js in a serviceworker?

### DISCUSSION THREAD

**Discussion by Ali-Javanmardi** [keycloak/keycloak#12739] [Q&A]
I searched about how to keep access token in browser in a safe place. In some blogs I found service-worker should be safer than keeping  token in main JavaScript memory or localstorage or sessionstorage.
for example: 
    1- https://blog.ropnop.com/storing-tokens-in-browser/
or here auth0 suggest using a Web worker to keep token in memory:
    2- https://auth0.com/docs/secure/security-guidance/data-security/token-storage

When I tried to access keycloak-js in a service-worker I found keycloak-js has several access to document element and window element that are not accessible in service-worker.

In the sample code in above mentioned blog (first link), he acquired token in main JavaScript thread and then sent it to service-worker by a postMessage.

I think if I create a keycloak object in main JavaScript thread, it will be accessible to all even if I send token to service-worker. I can not destroy keycloak object after login as I need it to refresh token later.

so IMO it was the best if I could initial keycloak object inside service-worker and so it could remain separate from main JavaScript thread from beginning.

Is there any suggestion for my problem? Or is there any plan to make keycloak-js accessible from a service-worker?

**Comment by cm253** [comment]
I don't think it's easy to port the current Keycloak JS adapter to a service worker. I agree that it would be better to store the tokens in a service worker instance, because it would be more secure and also allow to use the tokens from multiple browser tabs (with the same origin policy) at the same time. Also see my comment in https://github.com/keycloak/keycloak/issues/14122#issuecomment-1348086269.

The Keycloak JS adapter uses the old `XMLHttpRequest` API to fetch the tokens, so it's also not possible to implement a service worker that acts as a proxy for the `fetch` API calls. I think with the current implementation of the Keycloak JS adapter it's not possible to do this, or at least you have to rewrite parts of it.

**Comment by alexanderhofstaetter** [comment]
We also want to store the tokens in a service worker for the exact same reason!
